specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+during initialization and cannot be modified until another initialization. It is assumed that the chain operator
+will not modify the gas paying token address unless they specifically decide to do so and appropriately handle collateralization.
+
+If the gas paying token address is set to the `ETHER_TOKEN_ADDRESS` then the system is configured to use `ether` as the
+native asset on L2. Note that if nothing is stored in state at the `GAS_PAYING_TOKEN_SLOT` then it is considered that
+`ether` is used to pay for gas and getters on smart contracts SHOULD return the `ETHER_TOKEN_ADDRESS`.
+
+When the gas paying token is configured in the `SystemConfig`, it kicks off a special deposit transaction that is from the
+`DEPOSITOR_ACCOUNT` and calls the `L1Block` contract so that it can store the gas paying token address and its number of decimals
+in the L2 system.
+
+![](../static/assets/custom-token-setup.png)
+
+## Contract Modifications
+
+### OptimismPortal
+
+The `OptimismPortal` is updated with a new interface specifically for depositing custom tokens.
+
+#### `depositERC20Transaction`
+
+The `depositERC20Transaction` is useful for sending custom gas tokens to L2. It is broken out into its own interface
+to maintain backwards compatibility, simplify the implementation and prevent footguns when trying to deposit an
+ERC20 native asset into a chain that uses `ether` as its native asset.
+
+```solidity
+function depositERC20Transaction(
+    address _to,
+    uint256 _mint,
+    uint256 _value,
+    uint64 _gasLimit,
+    bool _isCreation,
+    bytes memory _data
+) public;
+```
+
+#### `depositTransaction`
+
+The `depositTransaction` function is updated to revert if there is non-zero `msg.value` when the system uses a custom
+native token. This allows existing applications that do cross domain calls to continue functioning as is.
+
+### StandardBridge
+
+The `StandardBridge` contracts on both L1 and L2 MUST be aware of the custom gas token. The `ether` specific ABI on the
+`StandardBridge` is disabled when custom gas token is enabled.
+
+This includes the following methods:
+
+- `bridgeETH(uint32,bytes)`
+- `bridgeETHTo(address,uint32,bytes)`
+- `receive()`
+
+The following diagram shows the control flow for when a user attempts to send `ether` through the `StandardBridge`, either
+on L1 or L2.
+
+![](../static/assets/custom-token-bridge-revert.png)
+
+### CrossDomainMessenger
+
+The `CrossDomainMessenger` contracts on both L1 and L2 MUST not be used to send the native asset between domains.
+It is possible to support this in the future, but for now the feature is disabled due to the high probability of footguns.
+Using the `CrossDomainMessenger` with a custom gas token would involve modifying the `sendMessage` implementation.
+
+### L1Block
+
+The `L1Block` contract is upgraded with a permissioned function that allows it to set the `GAS_PAYING_TOKEN_SLOT`.
```

Comment:
Addressed this comment. Would prefer to keep the logic for handling not being able to modify the token on L1 only in case somebody uses the storage setter contract method, would allow consistency, no need to bork that